### PR TITLE
Fixes #13043 - Change Parameter inheritance to DockerParameter

### PR DESCRIPTION
**Pull request by orrabin**

### CODE REVIEW

File: app/controllers/api/v2/containers_controller.rb [hunk 1]
```diff
@@ -53,7 +53,7 @@ def show
           param :memory, String
           param :cpu_shares, :number
           param :cpu_set, String
-          param :environment_variables, Hash
+          param :environment_variables, Array, :desc => N_("Optional array of environment variables hashes")
```
Comment (dLobatog) [inline app/controllers/api/v2/containers_controller.rb]: A small example would help there

File: app/controllers/containers/steps_controller.rb [hunk 2]
```diff
@@ -50,8 +54,9 @@ def state_params
               when :environment
                 [:tty, :docker_container_wizard_state_id,
                  :attach_stdin, :attach_stdout, :attach_stderr,
-                 :exposed_ports_attributes => [], :environment_variables_attributes => [],
-                 :dns_attributes => []
+                 :exposed_ports_attributes => docker_parameters_permited_params,
```
Comment (dLobatog) [inline app/controllers/containers/steps_controller.rb]: :clap:

File: app/models/concerns/foreman_docker/parameter_validators.rb [hunk 3]
```diff
@@ -1,11 +1,34 @@
...
-    include ::ParameterValidators
 
-    def parameters_symbol
-      return :environment_variables if is_a? Container
-      super
+    included do
+      validate :validate_parameters_keys
+    end
+
+    def validate_parameters_keys
```
Comment (dLobatog) [inline app/models/concerns/foreman_docker/parameter_validators.rb]: If this is a uniqueness validation, rename to `validate_unique_parameter_keys`

File: app/models/concerns/foreman_docker/parameter_validators.rb [hunk 4]
```diff
@@ -1,11 +1,34 @@
...
+      parameters_symbol = [:environment_variables, :exposed_ports, :dns]
+      parameters_symbol.each do |param_symbol|
+        keys  = []
+        errors = false
+
+        self.send(param_symbol).each do |param|
```
Comment (dLobatog) [inline app/models/concerns/foreman_docker/parameter_validators.rb]: `public_send` just as a safety guard

File: app/models/concerns/foreman_docker/parameter_validators.rb [hunk 5]
```diff
@@ -1,11 +1,34 @@
...
+          errors = parameters_validation(keys, param)
+        end
+
+        self.errors[param_symbol] = _('Please ensure the following parameters are unique') if errors
+      end
+    end
+
+    def parameters_validation(keys, param)
```
Comment (dLobatog) [inline app/models/concerns/foreman_docker/parameter_validators.rb]: `def duplicate_key?`

File: app/models/docker_parameter.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
+class DockerParameter < ActiveRecord::Base
+  extend FriendlyId
+  friendly_id :key
+  include Parameterizable::ByIdName
+
+  validates_lengths_from_database
+
+  include Authorizable
+  validates :key, :presence => true, :no_whitespace => true
```
Comment (dLobatog) [inline app/models/docker_parameter.rb]: I wonder if all the validations in the concern could've been substituted by a validation of uniqueness with scope here?

File: app/models/docker_parameter.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+  scoped_search :on => :key, :complete_value => true
+
+  default_scope -> { order("docker_parameters.key") }
+
+  before_validation :strip_whitespaces
```
Comment (dLobatog) [inline app/models/docker_parameter.rb]: What does this do if there's already a `no_whitespace` validation?

File: db/migrate/20160605134652_move_parameters_to_docker_parameters.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+   end
+   delete("delete from parameters where type in ('EnvironmentVariable', 'ForemanDocker::Dns', 'ExposedPort')")
+  end
+
+  def down
+    docker_params = select_all("select key, value, reference_id, type from docker_parameters where type in ('EnvironmentVariable', 'ForemanDocker::Dns', 'ExposedPort')")
+    docker_params.each do |param|
+      Parameter.create(:key => param['name'], :value => param['value'], :reference_id => param['reference_id'], :type => param['type'])
+    end
+    delete("delete from docker_parameters where type in ('EnvironmentVariable', 'ForemanDocker::Dns', 'ExposedPort')")
```
Comment (dLobatog) [inline db/migrate/20160605134652_move_parameters_to_docker_parameters.rb]: Why not using Rails SQL helpers instead which are portable between db adapters?

File: db/migrate/20160605134652_move_parameters_to_docker_parameters.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
+class MoveParametersToDockerParameters < ActiveRecord::Migration
+  def up
+  # All the  DockerContainerWizardStates::PARAMETER are temporary for the wizard step so no need to keep them
+   docker_params = select_all("select name, value, reference_id, type from parameters where type in ('EnvironmentVariable', 'ForemanDocker::Dns', 'ExposedPort')")
+   docker_params.each do |param|
+     DockerParameter.create(:key => param['name'], :value => param['value'], :reference_id => param['reference_id'], :type => param['type'])
```
Comment (dLobatog) [inline db/migrate/20160605134652_move_parameters_to_docker_parameters.rb]: Best to use a fake class w/o constraints rather than the real thing and risk failing the migration and potentially breaking an upgrade- see https://github.com/theforeman/foreman/blob/develop/db/migrate/20160516070529_divide_lookup_key_permissions.rb#L2 for an example

File: app/models/docker_parameter.rb (hunk 7 above)
Comment (orrabin) [inline app/models/docker_parameter.rb]: The value is also striped, I can remove the key from there.

File: db/migrate/20160605134652_move_parameters_to_docker_parameters.rb (hunk 9 above)
Comment (orrabin) [inline db/migrate/20160605134652_move_parameters_to_docker_parameters.rb]: I thought it was a problem to use fake objects here due to: http://projects.theforeman.org/issues/14983

**Comment by orrabin** [comment]
@dLobatog I fixed all the comments except the migration. I'd like to fix it but I'm not sure if the fake objects will help there because of the sti.

File: db/migrate/20160605134652_move_parameters_to_docker_parameters.rb (hunk 9 above)
Comment (dLobatog) [inline db/migrate/20160605134652_move_parameters_to_docker_parameters.rb]: `DockerParameter` is not part of the Foreman::STI, why would it be?

File: db/migrate/20160605134652_move_parameters_to_docker_parameters.rb (hunk 8 above)
Comment (dLobatog) [inline db/migrate/20160605134652_move_parameters_to_docker_parameters.rb]: Unfixed

File: test/functionals/api/v2/containers_controller_test.rb [hunk 10]
```diff
@@ -109,6 +109,21 @@ class ContainersControllerTest < ActionController::TestCase
                                           :tag => tag }
             assert_response :created
           end
+
+          test 'creates a container with correct params' do
```
Comment (dLobatog) [inline test/functionals/api/v2/containers_controller_test.rb]: Can you add another test for unwrapped parameters (by that I mean `post :create, { :compute_resource_id => ....}` without the top-level `:container =>` key)? 

It won't work right away, you'd have to add the equivalent of this `https://github.com/theforeman/foreman/pull/3864/files#diff-b167f4d1b1a23df4b43576b2dfacc1c2R7`  to the controller

File: app/models/docker_parameter.rb (hunk 7 above)
Comment (dLobatog) [inline app/models/docker_parameter.rb]: I mean, why not adding `validates :value, :presence => true,  :no_whitespace => true` too? Isn't it the same but reusing the code in the `no_whitespace` validation instead of writing your own `strip_whitespaces`?

File: app/models/docker_parameter.rb (hunk 6 above)
Comment (dLobatog) [inline app/models/docker_parameter.rb]: Can you add a test for that?

File: db/migrate/20160605134652_move_parameters_to_docker_parameters.rb (hunk 9 above)
Comment (orrabin) [inline db/migrate/20160605134652_move_parameters_to_docker_parameters.rb]: Becase EnvironmentVariable, ExposedPort and ForemanDocker::Dns exist and I had issues with trying to add FakeLookupKey in migrations. I'll switch it to a fake object and see if it does work in this case.

File: app/models/docker_parameter.rb (hunk 7 above)
Comment (orrabin) [inline app/models/docker_parameter.rb]: Because whitespace is allowed in value, it's just to remove trailing and leading whitespace.
This was in the original Parameters, if it's not needed I can remove it but then the values '  22' and '22' will not be considered identical.

File: test/functionals/api/v2/containers_controller_test.rb (hunk 10 above)
Comment (dLobatog) [inline test/functionals/api/v2/containers_controller_test.rb]: Will do in another PR - unrelated to this issue

File: app/models/docker_parameter.rb (hunk 6 above)
Comment (orrabin) [inline app/models/docker_parameter.rb]: I added a test

File: db/migrate/20160605134652_move_parameters_to_docker_parameters.rb [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+  class FakeParameter < ActiveRecord::Base
+    self.table_name = 'parameters'
+  end
+
+  def up
+  # All the  DockerContainerWizardStates::PARAMETER are temporary for the wizard step so no need to keep them
+   docker_params = FakeParameter.unscoped.where(:type => ['EnvironmentVariable', 'ForemanDocker::Dns', 'ExposedPort'])
+   docker_params.each do |param|
+     DockerParameter.create(:key => param['name'], :value => param['value'], :reference_id => param['reference_id'], :type => param['type'])
```
Comment (orrabin) [inline db/migrate/20160605134652_move_parameters_to_docker_parameters.rb]: I can't use the fake object here but used it to get rid of the `select`.

Comment (orrabin) [inline db/migrate/20160605134652_move_parameters_to_docker_parameters.rb]: I can't use the fake object here but used it to get rid of the `select_all`.

File: app/models/docker_parameter.rb (hunk 7 above)
Comment (dLobatog) [inline app/models/docker_parameter.rb]: Duh - I didn't realize it was doing it on value until now...

**Comment by dLobatog** [comment]
@orrabin Thanks! I'll test it and hopefully we can merge it after that

**Comment by dLobatog** [comment]
I'll fix http://ci.theforeman.org/job/test_plugin_matrix/2015/#showFailuresLink and merge this one afterwards. It works well locally and passes tests (except these failures). Thanks @orrabin
